components/style/sharing/mod.rs

```diff
@@ -109,6 +218,103 @@ pub struct StyleSharingCandidateCache<E: TElement> {
     cache: LRUCache<StyleSharingCandidate<E>>,
 }
 
+/// A struct we use as a simple cache of stuff we use to test against different
+/// candidates.
+#[derive(Debug)]
+pub struct ElementSharingChecker<E: TElement> {
+    element: E,
+    cache: CachedStyleSharingData,
+}
+
+impl<E: TElement> ElementSharingChecker<E> {
+    /// Creates a new empty checker for `element`.
+    pub fn new(element: E) -> Self {
+        Self {
+            element: element,
+            cache: CachedStyleSharingData::new(),
+        }
+    }
+
+    /// Consumes this checker and obtains the cached data.
+    pub fn take_cache(self) -> CachedStyleSharingData {
+        self.cache
+    }
+
+    /// Get the presentational hints affecting this element.
+    fn pres_hints(&mut self) -> &[ApplicableDeclarationBlock] {
+        self.cache.pres_hints(self.element)
+    }
+
+    /// Get the class list of this element.
+    fn class_list(&mut self) -> &[Atom] {
+        self.cache.class_list(self.element)
+    }
+
+    /// Computes the revalidation results if needed, and returns it.
+    fn revalidation_match_results(
+        &mut self,
+        stylist: &Stylist,
+        bloom: &BloomFilter,
+        selector_flags_map: &mut SelectorFlagsMap<E>
+    ) -> &BitVec {
+        let element = self.element;
+        // It's important to set the selector flags. Otherwise, if we succeed in
+        // sharing the style, we may not set the slow selector flags for the
+        // right elements (which may not necessarily be |element|), causing
+        // missed restyles after future DOM mutations.
+        //
+        // Gecko's test_bug534804.html exercises this. A minimal testcase is:
+        // <style> #e:empty + span { ... } </style>
+        // <span id="e">
+        //   <span></span>
+        // </span>
+        // <span></span>
+        //
+        // The style sharing cache will get a hit for the second span. When the
+        // child span is subsequently removed from the DOM, missing selector
+        // flags would cause us to miss the restyle on the second span.
+        let mut set_selector_flags = |el: &E, flags: ElementSelectorFlags| {
+            element.apply_selector_flags(selector_flags_map, el, flags);
+        };
+
+        self.cache.revalidation_match_results(self.element,
+                                              stylist,
+                                              bloom,
+                                              &mut set_selector_flags)
+    }
+
+    /// Attempts to share a style with another node. This method is unsafe
+    /// because it depends on the `style_sharing_candidate_cache` having only
+    /// live nodes in it, and we have no way to guarantee that at the type
+    /// system level yet.
+    pub unsafe fn share_style_if_possible(
```

Comment:
This comment and the associated unsafety are bogus now that we we stored this all in ScopedTLS. Feel free to fix.